scripts/calculateRevenue/protocols/beefy.ts

```diff
@@ -1,11 +1,100 @@
-import { RevenueResult } from '../../types'
+import { RevenueResult, NetworkId } from '../../types'
 import { fetchWithTimeout } from '../../utils/fetchWithTimeout'
-
+import { getViemPublicClient, getStrategyContract } from './utils/viem'
+import { Address } from 'viem'
 export type BeefyVaultTvlData = [string, number]
 
 const BEEFY_API_URL = 'https://databarn.beefy.com/api/v1/beefy'
+const DEFI_LLAMA_API_URL = 'https://coins.llama.fi'
+
 const ONE_WEEK = 7 * 24 * 60 * 60 * 1000
 
+const NETWORK_ID_TO_DEFI_LLAMA_CHAIN: Partial<{
+  [networkId in NetworkId]: string // eslint-disable-line @typescript-eslint/no-unused-vars
+}> = {
+  [NetworkId['ethereum-mainnet']]: 'Ethereum',
+  [NetworkId['arbitrum-one']]: 'Arbitrum',
+  [NetworkId['op-mainnet']]: 'Optimism',
+  [NetworkId['celo-mainnet']]: 'Celo',
+  [NetworkId['polygon-pos-mainnet']]: 'Polygon',
+  [NetworkId['base-mainnet']]: 'base',
+}
+
+export interface BlockTimestampData {
+  height: number
+  timestamp: number
+}
+
+interface FeeEvent {
+  beefyFee: number | bigint
+  timestamp: Date
+}
+
+// TODO: Memoize this function so it's not repeated for every user address
+/**
+ * Uses the DefiLlama API to fetch the block number nearest a given timestamp
+ */
+export async function getNearestBlock(
+  networkId: NetworkId,
+  timestamp: Date,
+): Promise<number> {
+  const unixTimestamp = Math.floor(timestamp.getTime() / 1000)
+  const defiLlamaChain = NETWORK_ID_TO_DEFI_LLAMA_CHAIN[networkId]
+
+  const response = await fetchWithTimeout(
+    `${DEFI_LLAMA_API_URL}/block/${defiLlamaChain}/${unixTimestamp}`,
+  )
+  if (!response.ok) {
+    console.log(
+      `${DEFI_LLAMA_API_URL}/block/${defiLlamaChain}/${unixTimestamp}`,
+    )
+    throw new Error(
+      `Error while fetching block timestamp from DefiLlama: ${response}`,
+    )
+  }
+  const blockTimestampData = (await response.json()) as BlockTimestampData
+  return blockTimestampData.height
+}
+
+// TODO: Memoize this function so it's not repeated for every user address
+/**
+ * For a given vault, fetches the record of all ChargedFee events emitted in a given timeframe
+ */
+export async function fetchFeeEvents(
+  vaultAddress: Address,
+  networkId: NetworkId,
+  startTimestamp: Date,
+  endTimestamp: Date,
+): Promise<FeeEvent[]> {
+  const client = getViemPublicClient(networkId)
+  const strategyContract = await getStrategyContract(vaultAddress, networkId)
+
+  const startBlock = await getNearestBlock(networkId, startTimestamp)
+  const endBlock = await getNearestBlock(networkId, endTimestamp)
+  const blocksPer = 10000
+  let currentBlock = startBlock
+
+  const feeEvents: FeeEvent[] = []
+  while (currentBlock < endBlock) {
+    const toBlock = Math.min(currentBlock + blocksPer, endBlock)
```

Comment:
Does this double-count the toBlock?  Like, imagine "blocksPer" is 1:

currentBlock = 0
toBlock = 1
getEvents({ fromBlock: currentBlock, toBlock }) ---> returns events from 0, and 1

currentBlock = 1
toBlock = 2
getEvents({ fromBlock: currentBlock, toBlock }) ---> returns events from 1, and 2

?

Reply:
Wow, I just assumed that the end range would be exclusive, but in fact it appears that both limits are _inclusive_, according to [here](https://ethereum.stackexchange.com/questions/78620/is-the-range-of-a-filterquery-inclusive-up-to-toblock). That's kind of annoying -- thanks for the callout.